wasm/src/lsp/parse.rs

```diff
@@ -126,3 +144,660 @@ fn test_doc_vec_byte_iter_0() {
+    }
+
+    // We ran into ( early.
+    if current_char == b'(' {
+        return true;
+    }
+
+    // Reverse past spaces.
+    while let Some((ch, p)) = lines.get_prev_position(&pos_walk) {
+        current_char = ch;
+
+        if !ch.is_ascii_whitespace() {
+            break;
+        }
+
+        pos_walk = p;
+    }
+
+    current_char == b'('
+}
+
+#[cfg(test)]
+fn make_simple_test_doc_data_from_lines(filename: &str, lines: &[&str]) -> DocData {
+    DocData {
+        fullname: filename.to_string(),
+        text: lines.iter().copied().map(vec_ref).collect(),
+        version: 1,
+        comments: Default::default()
+    }
+}
+
+#[test]
+fn test_is_first_in_list_simple() {
+    let simple_doc = make_simple_test_doc_data_from_lines("test.clsp", &[
+        "(hi there)"
+    ]);
+    assert!(is_first_in_list(&simple_doc, &Position { line: 0, character: 2 }));
+}
+
+#[test]
+fn test_is_first_in_list_with_spaces() {
+    let simple_doc = make_simple_test_doc_data_from_lines("test.clsp", &[
+        "(    hi there)"
+    ]);
+    assert!(is_first_in_list(&simple_doc, &Position { line: 0, character: 6 }));
+}
+
+#[test]
+fn test_is_first_in_list_diff_line() {
+    let simple_doc = make_simple_test_doc_data_from_lines("test.clsp", &[
+        "(",
+        "  hi there",
+        ")"
+    ]);
+    assert!(is_first_in_list(&simple_doc, &Position { line: 1, character: 2 }));
+}
+
+#[test]
+fn test_is_first_in_list_one_after() {
+    let simple_doc = make_simple_test_doc_data_from_lines("test.clsp", &[
+        "(hi there)"
+    ]);
+    assert!(is_first_in_list(&simple_doc, &Position { line: 0, character: 3 }));
+}
+
+#[test]
+fn test_not_is_first_in_list_next_word() {
+    let simple_doc = make_simple_test_doc_data_from_lines("test.clsp", &[
+        "(hi there)"
+    ]);
+    assert!(!is_first_in_list(&simple_doc, &Position { line: 0, character: 4 }));
+}
+
+// Given a position, return the identifier at that position.  Relies on find_ident.
+pub fn get_positional_text(lines: &DocData, position: &Position) -> Option<Vec<u8>> {
+    let pl = position.line as usize;
+    if pl < lines.text.len() {
+        let line = lines.text[pl].clone();
+        find_ident(line, position.character)
+    } else {
+        None
+    }
+}
+
+#[test]
+fn test_get_positional_text() {
+    let simple_doc = make_simple_test_doc_data_from_lines("test.clsp", &[
+        "(",
+        "  hi there",
+        ")"
+    ]);
+    assert_eq!(
+        get_positional_text(&simple_doc, &Position { line: 1, character: 3 }),
+        Some(b"hi".to_vec())
+    );
+}
+
+// Determine whether the character is an identifier char.
+pub fn is_identifier(v: &[u8]) -> bool {
+    v.iter().copied().all(is_identifier_char)
+}
+
+#[test]
+fn test_is_identifier() {
+    assert!(is_identifier(b"test"));
+}
+
+#[test]
+fn test_is_not_identifier() {
+    assert!(!is_identifier(b"; comment instead"));
+}
+
+fn make_inner_function_scopes(scopes: &mut Vec<ParseScope>, body: &BodyForm) {
```

Comment:
`cargo build` warned that this function is never used.
This should be made public or removed.

Reply:
Fairly certain I wrote tests for this.  Build will warn if things aren't used in the mode it's building in.
I can temporarily mark it cfg(test), but I'm building this out a part at a time to satisfy having a small amount of code at a time so not everything will be used yet.

Reply:
OK understood.